src/frontend/src/optimizer/plan_node/logical_source.rs

```diff
@@ -152,6 +159,50 @@ impl LogicalSource {
         Ok(Some(exprs))
     }
 
+    fn rewrite_new_s3_plan(&self) -> Result<PlanRef> {
+        let logical_source = generic::Source {
+            catalog: self.core.catalog.clone(),
+            column_catalog: vec![ColumnCatalog {
+                column_desc: ColumnDesc::from_field_with_column_id(
+                    &Field {
+                        name: "filename".to_string(),
+                        data_type: DataType::Varchar,
+                        sub_fields: vec![],
+                        type_name: "".to_string(),
+                    },
+                    0,
+                ),
+                is_hidden: false,
+            }],
+            row_id_index: None,
+            gen_row_id: false,
+            ..self.core.clone()
+        };
+        let mut new_s3_plan: PlanRef = StreamSource {
+            base: PlanBase::new_stream_with_logical(
+                &logical_source,
+                Distribution::Single,
+                logical_source
+                    .catalog
+                    .as_ref()
+                    .map_or(true, |s| s.append_only),
+                false,
+                FixedBitSet::with_capacity(logical_source.column_catalog.len()),
+            ),
+            logical: logical_source,
+        }
+        .into();
+        new_s3_plan = RequiredDist::shard_by_key(1, &[0])
+            .enforce_if_not_satisfies(new_s3_plan, &Order::any())?;
```

Comment:
@st1page any wrong with this part? I got a panic when explaining the query `explain create table t (v1 int, v2 varchar ) with ( connector = 's3_v2') format plain encode json ;`

Reply:
I will take a look